### PR TITLE
Refactor device module 368

### PR DESCRIPTION
**Pull request by guenthermi**
This is a draft to address issue #368. There is still an issue which occurs if specific paddle models (e.g. LSTM models) are assigned to a device with the paddle.fluid.dygraph.layers.to function.

### CODE REVIEW

File: finetuner/tailor/keras/__init__.py [hunk 1]
```diff
@@ -123,3 +123,6 @@ def to_embedding_model(
             model = tf.keras.Model(model.input, x)
 
         return model
+
+    def _set_device(self, device: str) -> None:
+        ...
```
[the remark this answers is not in the document]
Comment (guenthermi) [reply finetuner/tailor/keras/__init__.py:128]: In keras it is not possible to assign the model directly to a division. Instead one can use the context manager interface of the device object itself. Therefore, it is not necessary to set `self._device`.

File: finetuner/tailor/paddle/__init__.py [hunk 2]
```diff
@@ -202,6 +205,12 @@ def to_embedding_model(
 
         return model
 
+    def _set_device(self, device: str) -> None:
+        self._device = get_device_paddle(device)
+        for param in self._model.parameters():
+            if str(param.place) == 'CUDAPlace(0)':
+                raise DeviceError('Paddle models are expected to be placed on the CPU')
```
[the remark this answers is not in the document]
Comment (guenthermi) [reply finetuner/tailor/paddle/__init__.py:212]: It causes errors for LSTM models and maybe this happens for other models as well. Therefore, we decided to not support this.

**Review by bwanglzu — APPROVED** [review]
LGTM!

File: finetuner/tailor/keras/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmastrapas) [reply finetuner/tailor/keras/__init__.py:128]: ok got it

File: finetuner/tailor/paddle/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gmastrapas) [reply finetuner/tailor/paddle/__init__.py:212]: so no GPU support for tailoring but we support GPU for finetuning?

File: finetuner/device.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    import torch
+
+    # translate our own alias into framework-compatible ones
+    return torch.device(device)
+
+
+def to_device_pytorch(
+    inputs: Union[AnyTensor, Mapping[str, AnyTensor], Sequence[AnyTensor]], device
+) -> Union[AnyTensor, Dict[str, AnyTensor], List[AnyTensor]]:
+    """Map various of input type to device.
```
Comment (gmastrapas) [inline finetuner/device.py:20]: various input types

File: finetuner/device.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    else:
+        raise ValueError(
+            f'Device {device} not recognized, only "cuda" and "cpu" are accepted'
+        )
+
+
+def to_device_paddle(
+    inputs: Union[AnyTensor, Mapping[str, AnyTensor], Sequence[AnyTensor]], device
+) -> Union[AnyTensor, Dict[str, AnyTensor], List[AnyTensor]]:
+    """Map various of input type to device.
```
Comment (gmastrapas) [inline finetuner/device.py:57]: various input types

File: finetuner/tailor/paddle/__init__.py (hunk 2 above)
Comment (gmastrapas) [inline finetuner/tailor/paddle/__init__.py:212]: I would prefer a more informative message here, like
```
Tailoring a PaddlePaddle model placed on the GPU is currently not supported. Try placing to CPU instead.
```

**Review by gmastrapas — CHANGES_REQUESTED** [review]

**Review by gmastrapas — APPROVED** [review]
LGTM!

**Pull request merged**